### ISSUE TITLE
AppArmor profiles for NSM 

### ISSUE THREAD

**Issue by richardstone** [networkservicemesh/deployments-k8s#6580]
Hello!

I’m trying to set up AppArmor profiles for NSM, but got stuck on forwarder-vpp. I’m setting up an empty profile in “complain” mode, so that it will log what is needed for it to run, but this way the forwarder is failing to start. That is strange as in complain mode, there is no restrictions at all.

Setting up the profile on the worker:
```
$ sudo vi /etc/apparmor.d/forwarder-vpp
include <tunables/global>
 
# vim:syntax=apparmor
# AppArmor policy for forwarder
# ###AUTHOR###
# ###COPYRIGHT###
# ###COMMENT###
# No template variables specified
 
 
/bin/forwarder flags=(complain) {
 
}
 
/usr/bin/vpp flags=(complain) {
 
}

$ sudo systemctl reload apparmor
```

After applying the profile, I check if it’s enabled:
```
$ sudo apparmor_status
2 profiles are in complain mode.
   /bin/forwarder
   /usr/bin/vpp
```

But then the forwarder is failing to start:
```
Jun 23 08:29:47.915 [INFO] [cmd:/bin/forwarder] [duration:1.518379ms] completed phase 1: get config from environment
Jun 23 08:29:47.915 [INFO] [cmd:/bin/forwarder] executing phase 2: run vpp and get a connection to it (time since start: 1.675556ms)
Jun 23 08:29:47.915 [INFO] [cmd:/bin/forwarder] local vpp is being used
Jun 23 08:29:47.915 [INFO] [cmd:/bin/forwarder] [duration:421.821µs] completed phase 2: run vpp and get a connection to it
Jun 23 08:29:47.915 [WARN] [cmd:/bin/forwarder] skipping phases 3-5: no PCI resources config
Jun 23 08:29:47.915 [WARN] [cmd:/bin/forwarder] SR-IOV is not enabled
Jun 23 08:29:47.916 [INFO] [Config:ReadConfig] unmarshalled Config: &{Interfaces:[&{Name:eth1 Matches:[&{LabelSelector[&{Via:eth1}]}]}]}
Jun 23 08:29:47.915 [INFO] [cmd:vpp] vpp[15]: buffer: numa[0] falling back to non-hugepage backed buffer pool (vlib_buffer_main_init_numa_alloc: buffer size (4032) is greater than page size (1))
Jun 23 08:29:47.915 [INFO] [cmd:vpp] vpp[15]: vlib_buffer_main_init: failed to allocate buffer pool(s
Jun 23 08:29:48.497 [ERRO] [cmd:/bin/forwarder] %!s(<nil>)
Jun 23 08:29:48.497 [WARN] [cmd:/bin/forwarder] Link init failed context canceled;	error setting up device eth1;	github.com/networkservicemesh/cmd-forwarder-vpp/internal/vppinit.InitLinks;		/build/internal/vppinit/links.go:53;	main.main;		/build/main.go:192;	runtime.main;		/usr/lib64/go/1.18/src/runtime/proc.go:250;	runtime.goexit;		/usr/lib64/go/1.18/src/runtime/asm_amd64.s:1571;	
Jun 23 08:29:48.497 [INFO] [cmd:/bin/forwarder] executing phase 6: retrieving svid, check spire agent logs if this is the last line you see (time since start: 583.978417ms)
Jun 23 08:29:48.498 [FATA] error getting x509 source: context canceled
```

**Comment by szvincze** [comment]
`/usr/bin/vpp` is the suspect because I removed it from the apparmor profile and the forwarder is working properly and I found only "ALLOWED" apparmor messages for forwarder in `dmesg`.

**Comment by denis-tingaikin** [comment]
> I have started looking at it.

@szvincze Cool! Let us know if we need also have a look into this .

**Comment by szvincze** [comment]
Finally I found the way how to fix this issue.

It seems there is a problem when NSM components spawn additional processes, because it hits a disconnected path denial. It is the same with the probes that call different processes than the original `ENTRYPOINT` of the image.
(Note: `SPIRE` uses the same process for health-checks in the probes. This is why it works smoothly with `apparmor`.)

The Linux kernel does not force opening of handles to include the leading `/`, which `apparmor` has issues with.
By default `apparmor` rejects new accesses to disconnected paths and reports back the path-name without a leading `/`.
Unfortunately this breaks some NSM components. If a profile must allow for mediation of disconnected paths then the profile flag `attach_disconnected` shall be used. This prepend a leading `/` to the reported name and `apparmor` forces the disconnected objects to the namespace of the process and mediate them as though they are part of the namespace.

Here is the flag added to the example in the first comment:
```
/bin/forwarder flags=(complain, attach_disconnected) {
 
}
```
So, use this flag in the profile where needed.

Additionally I found that some of the paths are symlinks to others, like `/bin` to `/usr/bin` in case of the `forwarder-vpp` image and `apparmor` only enforces on real-paths. So, I had to use `/usr/bin/forwarder` in the profile.
Therefore always check the output of `aa-status` if it shows the processes defined in the profile as in the desired mode, like below:
```
9 processes are in complain mode.
   /opt/spire/bin/spire-agent (12572) 
   /opt/spire/bin/spire-agent (12732) 
   /opt/spire/bin/spire-server (11765) 
   /usr/bin/forwarder (100655) 
   /usr/bin/forwarder (100675) 
   /usr/bin/vpp (100760) /usr/bin/forwarder//usr/bin/vpp
   /usr/bin/vpp (100814) /usr/bin/forwarder//usr/bin/vpp
   /bin/nsmgr (100566) nsmgr
   /bin/nsmgr (100603) nsmgr
```
